### PR TITLE
FBI Check 2008-2017 data gets cleaned correctly

### PR DESCRIPTION
**Pull request by hanlu09205**
This change is [Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392)

### CODE REVIEW

**Review by pradh — APPROVED** [review]
nit: prefix FBI in PR title :)


Is there any documentation or something on why there are duplicates?

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392#-:-MSZuNbH2DtRelg2MnV9:bor42dy)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @hanlu09205)

___
*[scripts/fbi/crime/preprocess.py, line 212 at r3](https://reviewable.io/reviews/datacommonsorg/data/392#-MSZuYkSCL4WB1nvcI4M:-MSZuYkSCL4WB1nvcI4N:b-q2gpxk) ([raw file](https://github.com/datacommonsorg/data/blob/807cbc094dc7b91b2455623dcd4ba836b2b95f8d/scripts/fbi/crime/preprocess.py#L212)):*
<details><summary><i alt="summary">Quoted 81 lines of code…</i></summary>

> ```
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2680.0':
>         return True
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2753.0':
>         return True
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '516.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2694.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2762.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25771.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '509.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2743.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2784.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25599.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2769.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2806.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '515.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2778.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2808.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25583.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '518.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2807.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2807.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25662.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '521.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2826.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2813.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25558.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Newport' and field[_POPULATION_INDEX] == '1579.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '521.0':
>         return True
>     # 2016 CSV has footnote that we want to skip here.
>     if year == '2016' and field[_STATE_INDEX] is None:
>         return True
> ```
</details>


Consider making it a tuple list?

[
   ('OREGON', '2011', 'Lebanon', '25558.0'),
   ('OREGON', '2011', 'Portland', '521.0'),
   ...
]

**Review by hanlu09205 — COMMENTED** [review]
Further investigation points that it's actually due to bad line split. Update logic.

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392#-:-MSeK5jy6zFJ7WQ-yPty:b9kjyay)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @hanlu09205)

File: scripts/fbi/crime/preprocess.py [hunk 1]
```diff
@@ -111,15 +147,15 @@ def _int_from_field(f):
...
-    return False
+# Trim 2016 state name.
+def _get_2016_state(state):
+    substring1 = ' - Metropolitan Counties'
+    if substring1 in state:
+        return state.replace(substring1, '')
+    substring2 = ' - Nonmetropolitan Counties'
+    if substring2 in state:
+        return state.replace(substring2, '')
+    return state
```
Comment (pradh) [inline scripts/fbi/crime/preprocess.py:158]: nit: I think you can just do

```
  state = state.replace(' - Metropolitan Counties', '')
  state = state.replace(' - Nonmetropolitan Counties', '')
  return state
```

Or all in one line...

File: scripts/fbi/crime/preprocess.py [hunk 2]
```diff
@@ -275,27 +340,48 @@ def update_and_calculate_crime_csv(geo_codes, crime_csv, calculated_csv):
...
 
 
+def create_tmcf_file(tmcf_file_path):
+    stat_vars = OUTPUT_COLUMNS[2:]
+    with open(tmcf_file_path, 'w', newline='') as f_out:
+        for i in range(len(stat_vars)):
+            f_out.write(
+                TEMPLATE_MCF_TEMPLATE.format_map({
+                    'index': i,
+                    'stat_var': OUTPUT_COLUMNS[2:][i]
```
Comment (pradh) [inline scripts/fbi/crime/preprocess.py:350]: nit: stat_vars[i]

File: scripts/fbi/crime/preprocess.py [hunk 3]
```diff
@@ -111,15 +147,15 @@ def _int_from_field(f):
...
-# When there is two entries for the same city, skip the one with wrong population data.
-def _should_skip_special_line(year, field):
-    if year == '2019' and field[_STATE_INDEX] == 'OREGON' and field[
-            _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2670.0':
-        return True
-    if year == '2019' and field[_STATE_INDEX] == 'OREGON' and field[
-            _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25959.0':
-        return True
-    return False
+# Trim 2016 state name.
```
Comment (pradh) [inline scripts/fbi/crime/preprocess.py:150]: What was the issue before that required custom population handling? Can we update the README which still talks about duplicate city population (https://github.com/datacommonsorg/data/blob/master/scripts/fbi/crime/README.md)

**Review by pradh — APPROVED** [review]

**Review by hanlu09205 — COMMENTED** [review]
Thanks for the review and help!

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392#-:-MSjXLUPESyNZm-Gpm-n:b-iyckwp)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @pradh)

___
*[scripts/fbi/crime/preprocess.py, line 212 at r3](https://reviewable.io/reviews/datacommonsorg/data/392#-MSZuYkSCL4WB1nvcI4M:-MSjWW7B7_5e3A5I4-P7:br6yrk6) ([raw file](https://github.com/datacommonsorg/data/blob/807cbc094dc7b91b2455623dcd4ba836b2b95f8d/scripts/fbi/crime/preprocess.py#L212)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2680.0':
>         return True
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2753.0':
>         return True
>     if year == '2018' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '516.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2694.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2762.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25771.0':
>         return True
>     if year == '2017' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '509.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2743.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2784.0':
>         return True
>     if year == '2015' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25599.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2769.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2806.0':
>         return True
>     if year == '2014' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '515.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2778.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2808.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25583.0':
>         return True
>     if year == '2013' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '518.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2807.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2807.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25662.0':
>         return True
>     if year == '2012' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '521.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Ashland' and field[_POPULATION_INDEX] == '2826.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Dallas' and field[_POPULATION_INDEX] == '2813.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Lebanon' and field[_POPULATION_INDEX] == '25558.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Newport' and field[_POPULATION_INDEX] == '1579.0':
>         return True
>     if year == '2011' and field[_STATE_INDEX] == 'OREGON' and field[
>             _CITY_INDEX] == 'Portland' and field[_POPULATION_INDEX] == '521.0':
>         return True
>     # 2016 CSV has footnote that we want to skip here.
>     if year == '2016' and field[_STATE_INDEX] is None:
>         return True
> ```

Consider making it a tuple list?

[
   ('OREGON', '2011', 'Lebanon', '25558.0'),
   ('OREGON', '2011', 'Portland', '521.0'),
   ...
]
</blockquote></details>

No longer needed, this function. 
___
*[scripts/fbi/crime/preprocess.py, line 150 at r5](https://reviewable.io/reviews/datacommonsorg/data/392#-MSZmvPvk6FBJIBX5ykI-r5-150:-MSjVSfsB2RcxN1jN9vy:b-boaho0) ([raw file](https://github.com/datacommonsorg/data/blob/a133c9b415f79098d556cd94c3f5ddb7db2ebbac/scripts/fbi/crime/preprocess.py#L150)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

What was the issue before that required custom population handling? Can we update the README which still talks about duplicate city population (https://github.com/datacommonsorg/data/blob/master/scripts/fbi/crime/README.md)
</blockquote></details>

# Split by comma and exclude comma from quotes in split
# For case like PENNSYLVANIA,"Abington Township, Montgomery County",55476.0,53.0,0.0,6.0,0,15.0,32.0,934.0,32.0,883.0,19.0,2.0
Previously, the split would break  "Abington Township, Montgomery County" causing that line to be dropped. 
And the city in PENNSYLVANIA would be marked as city in Oregon and cause duplication. 

README updated. 
___
*[scripts/fbi/crime/preprocess.py, line 158 at r5](https://reviewable.io/reviews/datacommonsorg/data/392#-MSZmvPvk6FBJIBX5ykI-r5-158:-MSjX5qt-NwJsULGjXr8:brxfia9) ([raw file](https://github.com/datacommonsorg/data/blob/a133c9b415f79098d556cd94c3f5ddb7db2ebbac/scripts/fbi/crime/preprocess.py#L158)):*
>  state = state.replace(' - Metropolitan Counties', '')
>   state = state.replace(' - Nonmetropolitan Counties', '')
>   return state
Done.
___
*[scripts/fbi/crime/preprocess.py, line 350 at r5](https://reviewable.io/reviews/datacommonsorg/data/392#-MSZmvPvk6FBJIBX5ykI-r5-350:-MSjXFl0EgBWV0No-CF3:b-6g5a53) ([raw file](https://github.com/datacommonsorg/data/blob/a133c9b415f79098d556cd94c3f5ddb7db2ebbac/scripts/fbi/crime/preprocess.py#L350)):*
> stat_vars[i]
Done.

File: scripts/fbi/crime/README.md [hunk 4]
```diff
@@ -15,6 +15,7 @@ and it relies on these statistic variables:
 - From 2013-2016, the FBI reported statistics for two different definitions of rape before fully transitioning to the current definition in 2017. We add a dummy column after it (so all years have two Rape columns).
 - 2016 FBI reported data is missing population. We blindly add one column with value 1. 
 - For duplicate city data, we remove the one with wrong population data (Google search city population). 
+When it happens, you will see error message like "duplicate city state". Add logic in clean_crime_file to ignore the wrong one.
```
Comment (pradh) [inline scripts/fbi/crime/README.md:18]: Do we even have a duplicate city anymore?

Comment (pradh) [inline scripts/fbi/crime/README.md:18]: `clean_crime_file()` function

**Review by hanlu09205 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @pradh)

___
*[scripts/fbi/crime/README.md, line 18 at r6](https://reviewable.io/reviews/datacommonsorg/data/392#-MSjY2A3vpOqW7MmusqR-r6-18:-MSjqFgIChA0wfOH7sYn:bd6l8vx) ([raw file](https://github.com/datacommonsorg/data/blob/b59cd86e49210680493f76504ae6b6c18c5fcbc0/scripts/fbi/crime/README.md#L18)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

`clean_crime_file()` function
</blockquote></details>

Done.

Not any more..

**Review by hanlu09205 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/392)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @pradh)

___
*[scripts/fbi/crime/README.md, line 18 at r6](https://reviewable.io/reviews/datacommonsorg/data/392#-MSjY2A3vpOqW7MmusqR-r6-18:-MSjx5_MDxZ7e_ZwMqZy:b-i2d92x) ([raw file](https://github.com/datacommonsorg/data/blob/b59cd86e49210680493f76504ae6b6c18c5fcbc0/scripts/fbi/crime/README.md#L18)):*
<details><summary><i>Previously, hanlu09205 wrote…</i></summary><blockquote>

Done.

Not any more..
</blockquote></details>

README reworded as below:
Right now, there is no duplicate city state. In case it happens in the future, you will see error message like "duplicate city state". Add logic in clean_crime_file() function to ignore the one with wrong population data. 

Submitting...

**Pull request merged**